source/common/common/stack_array.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#if !defined(WIN32)
+#include <alloca.h>
+
+#else
+#include <malloc.h>
+#endif
+
+#include <stddef.h>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+
+// This macro is intended to be used as a replacement for variable-length arrays.
+// Note that the StackArray wrapper object will be destructed and each element's
+// destructor will be called when it leaves scope. However, the memory containing
+// the array won't be deallocated until the function containing the macro returns
+#define STACK_ARRAY(var, type, num) StackArray<type> var(::alloca(sizeof(type) * num), num)
```

Comment:
Thinking about this, do we actually need this macro? Couldn't we just specify the template instantiation directly with type and num for each var and actually have `:alloca` be done as part of the constructor? Or does that now work because then the alloca would be cleaned up on constructor exit? If that's the case do you mind adding a small comment to that effect?

Reply:
That's correct, if we called `alloca` in the constructor, the memory would be freed when the constructor returns

Reply:
OK makes sense thanks.